README.md
# timezh - Chinese style time

> Golang库 - 时间格式中文化

[![Build Status](https://travis-ci.org/WindomZ/timezh.svg?branch=master)](https://travis-ci.org/WindomZ/timezh)
[![Coverage Status](https://coveralls.io/repos/github/WindomZ/timezh/badge.svg?branch=master)](https://coveralls.io/github/WindomZ/timezh?branch=master)
[![GoDoc](https://godoc.org/github.com/WindomZ/timezh?status.svg)](https://godoc.org/github.com/WindomZ/timezh)
[![Go Report Card](https://goreportcard.com/badge/github.com/WindomZ/timezh)](https://goreportcard.com/report/github.com/WindomZ/timezh)

扩展原生`time`快速实现时间格式中文化

## Install
```bash
go get github.com/WindomZ/timezh
```

### Roadmap
- [x] Format
  - [x] Format(string) string
  - [x] FormatMix(string) string
  - [x] FormatLayout(string) string
  - [x] FormatValue(string) string
- [x] Parse
  - [x] Parse(string, string) (Time, error)
  - [x] ParseLayout(string) string
  - [x] ParseValue(string) string

## Usage
|格式用途|原生`time`|引入`timezh`|
|:-----:|:-----:|:-----:|
|月份|January, Jan|一月|
|星期|Monday, Mon|星期一, 周一|
|上下午|PM, pm|下午|

`time`库**不支持**中文化：
```go
import "time"

time.Now().Format("2006年01月02日(January) PM3:04:05 Monday(Mon)")

>>> 2009年01月03日(January) PM6:15:05 Saturday(Sat)
```

`timezh`库**支持**中文化：
```go
import "github.com/WindomZ/timezh"

timezh.Now().Format("2006年01月02日(一月) 下午3:04:05 星期一(周一)")

>>> 2009年01月03日(一月) 下午6:15:05 星期六(周六)
```

`timezh`库**支持**中文解析：
```go
import "github.com/WindomZ/timezh"

timezh.Parse("2006年01月02日(一月) 下午3:04:05 星期一(周一)", "2009年01月03日(一月) 下午6:15:05 星期六(周六)")

time.Now().Format("2006年01月02日(January) PM3:04:05 Monday(Mon)")
timezh.Now().Format("2006年01月02日(一月) 下午3:04:05 星期一(周一)")

>>> 2009年01月03日(January) PM6:15:05 Saturday(Sat)
>>> 2009年01月03日(一月) 下午6:15:05 星期六(周六)
```

### Advanced
混杂中英文：
```go
import "github.com/WindomZ/timezh"

timezh.Now().FormatMix("2006年01月02日(January, 一月) 下午3:04:05PM 星期一(Mon, 周一)")

>>> 2009年01月03日(January, 一月) 下午6:15:05PM 星期六(Sat, 周六)
```

文本中文化：
```go
import "github.com/WindomZ/timezh"

timezh.FormatValue("2009年01月03日(January) 6:15:05PM 星期六(Sat)")

>>> 2009年01月03日(一月) 6:15:05下午 星期六(周六)
```

文本中文解析：
```go
import "github.com/WindomZ/timezh"

timezh.ParseValue("2009年01月03日(一月) 6:15:05下午 星期六(周六)")

>>> 2009年01月03日(Jan) 6:15:05PM Saturday(Sat)
```

## Benchmark
详见[BENCHMARK](BENCHMARK.md)

## Contributing
欢迎你Fork，提交PR，在[issues page](https://github.com/WindomZ/timezh/issues)汇报Bugs、提出新想法等，
我很乐意能一起参与。

如果你喜欢这个项目，可以点下 :star: 予以支持，谢谢！

## 许可
[MIT](https://github.com/WindomZ/timezh/blob/master/LICENSE)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## 1.0.0 - 2018-04-21
### Added
- First version release.

[Unreleased]: https://github.com/WindomZ/timezh/compare/v1.0.0...HEAD


time_test.go
package timezh

import (
	"testing"
	"time"

	"github.com/WindomZ/testify/assert"
)

func TestSince(t *testing.T) {
	assert.NotZero(t, Since(Now()))
	assert.True(t, Since(Now()) <= time.Millisecond)
}

func TestUntil(t *testing.T) {
	assert.NotZero(t, Until(Now()))
	assert.True(t, Until(Now()) <= time.Millisecond)
}

func TestUnix(t *testing.T) {
	assert.False(t, Unix(1, 0).IsZero())
}

func TestDate(t *testing.T) {
	assert.False(t, Date(1970, 1, 1, 0, 0, 0, 0, time.Local).IsZero())
}


parse_test.go
package timezh

import (
	"testing"
	"time"

	"github.com/WindomZ/testify/assert"
)

func TestParseLayout(t *testing.T) {
	layout1 := "January, Jan, 2006-01-02, 02-Jan-06, Monday, Mon, 15:04:05PM -0700, 15:04:05Z07:00PM MST"
	assert.Equal(t, layout1, ParseLayout(layout1))
	assert.Equal(t, layout1, ParseLayout(FormatLayout(layout1)))

	layout21 := "一月, 2006-01-02, 02-Jan-06, 星期一, 周一, Mon, 15:04:05下午 -0700, 15:04:05Z07:00PM MST"
	layout22 := "Jan, 2006-01-02, 02-Jan-06, Monday, Mon, Mon, 15:04:05PM -0700, 15:04:05Z07:00PM MST"
	assert.Equal(t, layout22, ParseLayout(layout21))
	assert.Equal(t, layout22, ParseLayout(FormatLayout(layout21)))
}

func TestParseValue(t *testing.T) {
	value11 := "一月, 2006-01-02, 02-Jan-06, 星期一, 周一, Mon, 15:04:05下午 -0700, 15:04:05Z07:00PM MST"
	value12 := "Jan, 2006-01-02, 02-Jan-06, Monday, Mon, Mon, 15:04:05PM -0700, 15:04:05Z07:00PM MST"
	assert.Equal(t, value12, ParseValue(value11))

	value21 := "二月, 2006-01-02, 02-Tue-06, 星期二, 周二, Tue, 15:04:05下午 -0700, 15:04:05Z07:00PM MST"
	value22 := "Feb, 2006-01-02, 02-Tue-06, Tuesday, Tue, Tue, 15:04:05PM -0700, 15:04:05Z07:00PM MST"
	assert.Equal(t, value22, ParseValue(value21))
}

func TestParse(t *testing.T) {
	testTime, _ := time.Parse("2006-01-02 15:04:05", "2009-01-03 18:15:05")

	layout1 := "Jan, 2006-01-02, 02-Jan-06, Mon, 15:04:05PM -0700, 15:04:05Z07:00PM MST"
	if result, err := Parse(layout1, testTime.Format(layout1)); assert.NoError(t, err) {
		assert.Equal(t, testTime.UnixNano(), result.UnixNano())
	}
	if result, err := Parse(layout1, T(testTime).Format(layout1)); assert.NoError(t, err) {
		assert.Equal(t, testTime.UnixNano(), result.UnixNano())
	}

	layout2 := "一月, 2006-01-02, 02-Jan-06, 星期一, 周一, Mon, 15:04:05下午 -0700, 15:04:05Z07:00PM MST"
	if result, err := Parse(layout2, testTime.Format(layout2)); assert.NoError(t, err) {
		assert.Equal(t, testTime.UnixNano(), result.UnixNano())
	}
	if result, err := Parse(layout2, T(testTime).Format(layout2)); assert.NoError(t, err) {
		assert.Equal(t, testTime.UnixNano(), result.UnixNano())
	}
}

func BenchmarkParse1(b *testing.B) {
	b.StopTimer()
	b.ReportAllocs()
	b.StartTimer()
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			time.Parse("2006年01月02日(January) PM3:04:05 Monday(Mon)",
				"2009年01月03日(Jan) PM6:15:05 Saturday(Sat)")
		}
	})
}

func BenchmarkParse2(b *testing.B) {
	b.StopTimer()
	b.ReportAllocs()
	b.StartTimer()
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			Parse("2006年01月02日(January) PM3:04:05 Monday(Mon)",
				"2009年01月03日(一月) 下午6:15:05 星期六(周六)")
		}
	})
}

func BenchmarkParse3(b *testing.B) {
	b.StopTimer()
	b.ReportAllocs()
	b.StartTimer()
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			time.Parse("2006年01月02日(January) PM3:04:05 Monday(Mon)",
				ParseValue("2009年01月03日(Jan) PM6:15:05 Saturday(Sat)"))
		}
	})
}

func BenchmarkParseValue(b *testing.B) {
	b.StopTimer()
	b.ReportAllocs()
	b.StartTimer()
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			ParseValue("2009年01月03日(一月) 6:15:05下午 星期六(周六)")
		}
	})
}


format_test.go
package timezh

import (
	"testing"
	"time"

	"github.com/WindomZ/testify/assert"
)

func TestTime_Format(t *testing.T) {
	testTime, _ := time.Parse("2006-01-02 15:04:05", "2009-01-03 18:15:05")

	assert.Equal(t, "2009-01-03(一月) 下午6:15:05 星期六(周六)",
		T(testTime).Format("2006-01-02(January) PM3:04:05 Monday(Mon)"))
	assert.Equal(t, "2009年01月03日(一月) 下午6:15:05 星期六(周六)",
		T(testTime).Format("2006年01月02日(January) PM3:04:05 Monday(Mon)"))
	assert.Equal(t, "2009年01月03日(一月) 下午6:15:05 星期六(周六)",
		T(testTime).Format("2006年01月02日(一月) 下午3:04:05 星期一(周一)"))
	assert.Equal(t, "2009年01月03日(一月, 一月) 下午6:15:05下午 星期六(周六, 周六)",
		T(testTime).Format("2006年01月02日(January, 一月) 下午3:04:05PM 星期一(Mon, 周一)"))
}

func TestTime_FormatMix(t *testing.T) {
	testTime, _ := time.Parse("2006-01-02 15:04:05", "2009-01-03 18:15:05")
	assert.Equal(t, "2009年01月03日(January, 一月) 下午6:15:05PM 星期六(Sat, 周六)",
		T(testTime).FormatMix("2006年01月02日(January, 一月) 下午3:04:05PM 星期一(Mon, 周一)"))

	testTime, _ = time.Parse("2006-01-02 15:04:05", "2009-01-04 02:15:05")
	assert.Equal(t, "2009年01月04日(January, 一月) 上午2:15:05AM 星期日(Sun, 周日)",
		T(testTime).FormatMix("2006年01月02日(January, 一月) 下午3:04:05PM 星期一(Mon, 周一)"))
}

func TestTime_Format_All(t *testing.T) {
	layout1 := "January, Jan, 2006-01-02, 02-Jan-06, Monday, Mon, 15:04:05PM -0700, 15:04:05Z07:00PM MST"
	assert.NotEqual(t, time.Now().Format(layout1), Now().Format(layout1))
	assert.NotEqual(t, Now().Format(layout1), Now().FormatMix(layout1))

	layout2 := "January, 一月, Jan, 2006-01-02, 02-一月-06, Monday, 星期一, Mon, 周一, 15:04:05PM -0700, 15:04:05Z07:00下午 MST"
	assert.NotEqual(t, layout2, FormatLayout(layout2))
	assert.NotEqual(t, time.Now().Format(layout2), Now().Format(layout2))
	assert.NotEqual(t, Now().Format(layout2), Now().FormatMix(layout2))

	layout3 := "January, 一月, Jan, 2006-01-02, 02-一月-06, Monday, 星期一, Mon, 周一, 15:04:05PM -0700, 15:04:05.999999999Z07:00下午 MST"
	assert.NotEqual(t, layout3, FormatLayout(layout3))
	assert.NotEqual(t, time.Now().Format(layout3), Now().Format(layout3))
	assert.NotEqual(t, Now().Format(layout3), Now().FormatMix(layout3))

	layout4 := "15:04:05PM party time"
	assert.Equal(t, layout4, FormatLayout(layout4))
	assert.NotEqual(t, time.Now().Format(layout4), Now().Format(layout4))
	assert.NotEqual(t, Now().Format(layout4), Now().FormatMix(layout4))

	layout5 := "all "
	for _, month := range shortMonthNames {
		layout5 += month + " "
	}
	for _, month := range longMonthNames {
		layout5 += month + " "
	}
	for _, day := range longDayNames {
		layout5 += day + " "
	}
	for _, day := range shortDayNames {
		layout5 += day + " "
	}
	assert.Equal(t, layout5, FormatLayout(layout5))
	assert.NotEqual(t, time.Now().Format(layout5), Now().Format(layout5))
	assert.NotEqual(t, Now().Format(layout5), Now().FormatMix(layout5))

	layout6 := "15:04:05PM 15:04:05pm 3:04:05PM 3:04:05pm"
	layout7 := "15:04:05PM 15:04:05pm 3:04:05PM 3:04:05pm 15:04:05下午 3:04:05下午"

	testTime, _ := time.Parse("2006-01-02 15:04:05", "2009-01-03 18:15:05")

	assert.NotEqual(t, layout6, FormatLayout(layout6))
	assert.NotEqual(t, testTime.Format(layout6), T(testTime).Format(layout6))
	assert.Equal(t, "18:15:05PM 18:15:05PM 6:15:05PM 6:15:05PM",
		T(testTime).FormatMix(layout6))

	assert.NotEqual(t, layout7, FormatLayout(layout7))
	assert.NotEqual(t, testTime.Format(layout7), T(testTime).Format(layout7))
	assert.Equal(t, "18:15:05PM 18:15:05PM 6:15:05PM 6:15:05PM 18:15:05下午 6:15:05下午",
		T(testTime).FormatMix(layout7))

	testTime, _ = time.Parse("2006-01-02 15:04:05", "2009-01-04 02:15:05")

	assert.NotEqual(t, layout6, FormatLayout(layout6))
	assert.NotEqual(t, testTime.Format(layout6), T(testTime).Format(layout6))
	assert.Equal(t, "02:15:05AM 02:15:05AM 2:15:05AM 2:15:05AM",
		T(testTime).FormatMix(layout6))

	assert.NotEqual(t, layout7, FormatLayout(layout7))
	assert.NotEqual(t, testTime.Format(layout7), T(testTime).Format(layout7))
	assert.Equal(t, "02:15:05AM 02:15:05AM 2:15:05AM 2:15:05AM 02:15:05上午 2:15:05上午",
		T(testTime).FormatMix(layout7))
}

func TestFormatLayout(t *testing.T) {
	layout1 := "January, Jan, 2006-01-02, 02-Jan-06, Monday, Mon, 15:04:05PM -0700, 15:04:05Z07:00PM MST"
	assert.Equal(t, layout1, FormatLayout(layout1))
}

func TestFormatValue(t *testing.T) {
	assert.Equal(t, "一月, 一月, 02-一月-06, 星期一, 周一, 15:04:05下午",
		FormatValue("January, Jan, 02-Jan-06, Monday, Mon, 15:04:05PM"))
}

func BenchmarkFormatLayout(b *testing.B) {
	b.StopTimer()
	b.ReportAllocs()
	b.StartTimer()
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			FormatLayout("2006年01月02日(January, 一月) 下午3:04:05PM 星期一(Mon, 周一)")
		}
	})
}

func BenchmarkFormatValue(b *testing.B) {
	b.StopTimer()
	b.ReportAllocs()
	b.StartTimer()
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			FormatLayout("January, Jan, 02-Jan-06, Monday, Mon, 15:04:05PM")
		}
	})
}

func BenchmarkTime_Format1(b *testing.B) {
	b.StopTimer()
	t := time.Now()
	b.ReportAllocs()
	b.StartTimer()
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			t.Format("2006年01月02日 January 3:04:05PM Mon")
		}
	})
}

func BenchmarkTime_Format2(b *testing.B) {
	b.StopTimer()
	t := Now()
	b.ReportAllocs()
	b.StartTimer()
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			t.Format("2006年01月02日 一月 3:04:05下午 星期一")
		}
	})
}

func BenchmarkTime_FormatMix(b *testing.B) {
	b.StopTimer()
	t := Now()
	t.FormatMix("2006年01月02日(January, 一月) 下午3:04:05PM 星期一(Mon, 周一)")
	b.ReportAllocs()
	b.StartTimer()
	b.RunParallel(func(pb *testing.PB) {
		for pb.Next() {
			t.FormatMix("2006年01月02日(January, 一月) 下午3:04:05PM 星期一(Mon, 周一)")
		}
	})
}


time.go
package timezh

import "time"

// Time same as time.Time, but support Chinese formatting.
type Time struct {
	time.Time
	f format
}

// T returns a Time instance from time.Time.
func T(t time.Time) Time {
	return Time{Time: t}
}

// Since returns the time elapsed since t.
// It is shorthand for time.Now().Sub(t).
func Since(t Time) time.Duration {
	return time.Now().Sub(t.Time)
}

// Until returns the duration until t.
// It is shorthand for t.Sub(time.Now()).
func Until(t Time) time.Duration {
	return t.Sub(time.Now())
}

// Now returns the current local time.
func Now() Time {
	return T(time.Now())
}

// Unix returns the local Time corresponding to the given Unix time,
// sec seconds and nsec nanoseconds since January 1, 1970 UTC.
// It is valid to pass nsec outside the range [0, 999999999].
// Not all sec values have a corresponding time value. One such
// value is 1<<63-1 (the largest int64 value).
func Unix(sec int64, nsec int64) Time {
	return T(time.Unix(sec, nsec))
}

// Date returns the Time corresponding to
//	yyyy-mm-dd hh:mm:ss + nsec nanoseconds
// in the appropriate zone for that time in the given location.
//
// The month, day, hour, min, sec, and nsec values may be outside
// their usual ranges and will be normalized during the conversion.
// For example, October 32 converts to November 1.
//
// A daylight savings time transition skips or repeats times.
// For example, in the United States, March 13, 2011 2:15am never occurred,
// while November 6, 2011 1:15am occurred twice. In such cases, the
// choice of time zone, and therefore the time, is not well-defined.
// Date returns a time that is correct in one of the two zones involved
// in the transition, but it does not guarantee which.
//
// Date panics if loc is nil.
func Date(year int, month time.Month, day, hour, min, sec, nsec int, loc *time.Location) Time {
	return T(time.Date(year, month, day, hour, min, sec, nsec, loc))
}


parse.go
package timezh

import "time"

// Parse parses a formatted string and returns the time value it represents.
// The layout defines the format by showing how the reference time,
// defined to be
//	星期一 一月 2 下午15:04:05 -0700 MST 2006
// Same as time.Parse, but support Chinese string.
func Parse(layout, value string) (t Time, err error) {
	t.Time, err = time.Parse(t.parseLayout(layout), t.parseValue(value))
	return
}

// ParseLayout returns a standard textual formatted layout.
// Such as
//	星期一 一月 2 下午15:04:05 -0700 MST 2006
// returns
//	Mon Jan 2 PM15:04:05 -0700 MST 2006
func ParseLayout(layout string) string {
	return new(Time).parseLayout(layout)
}

func (t *Time) parseLayout(layout string) string {
	t.f.reset()
	j := 0
	for i, r := range layout {
		if i < j {
			continue
		}
		switch r {
		case '星':
			if len(layout) >= i+9 && layout[i:i+9] == "星期一" {
				j = i + 9
				t.f.buf.WriteString("Monday")
			}
		case '周':
			if len(layout) >= i+6 && layout[i:i+6] == "周一" {
				j = i + 6
				t.f.buf.WriteString("Mon")
			}
		case '一':
			if len(layout) >= i+6 && layout[i:i+6] == "一月" {
				j = i + 6
				t.f.buf.WriteString("Jan")
			}
		case '下':
			if len(layout) >= i+6 && layout[i:i+6] == "下午" {
				j = i + 6
				t.f.buf.WriteString("PM")
			}
		}
		if j <= i {
			t.f.buf.WriteRune(r)
		}
	}

	return t.f.buf.String()
}

// ParseValue returns a standard textual from
// a textual representation of the time value formatted.
// Such as
//	星期二 二月 2 下午15:04:05 -0700 MST 2006
// returns
//	Tue Feb 2 PM15:04:05 -0700 MST 2006
func ParseValue(value string) string {
	return new(Time).parseValue(value)
}

func (t *Time) parseValue(value string) string {
	t.f.reset()
	j := 0
	for i, r := range value {
		if i < j {
			continue
		}
		switch r {
		case '星':
			if len(value) >= i+9 {
				s := value[i : i+9]
				for k, day := range longDayMap {
					if s == day {
						j = i + 9
						t.f.buf.WriteString(longDayNames[k])
						break
					}
				}
			}
		case '周':
			if len(value) >= i+6 {
				s := value[i : i+6]
				for k, day := range shortDayMap {
					if s == day {
						j = i + 6
						t.f.buf.WriteString(shortDayNames[k])
						break
					}
				}
			}
		case '一', '二', '三', '四', '五', '六', '七', '八', '九', '十':
			if len(value) >= i+6 {
				s := value[i : i+6]
				for k, day := range shortMonthMap {
					if s == day {
						j = i + 6
						t.f.buf.WriteString(shortMonthNames[k])
						break
					}
				}
			}
		case '上', '下':
			if len(value) >= i+6 {
				s := value[i : i+6]
				for k, day := range longPMMap {
					if s == day {
						j = i + 6
						t.f.buf.WriteString(longPMNames[k])
						break
					}
				}
			}
		}
		if j <= i {
			t.f.buf.WriteRune(r)
		}
	}

	return t.f.buf.String()
}


format.go
package timezh

import "bytes"

type format struct {
	mixed      bool
	day        uint8
	dayPoint   uint8
	month      uint8
	monthPoint uint8
	pm         uint8
	pmPoint    uint8
	buf        *bytes.Buffer
}

func (f *format) init(s string) {
	f.day = 0
	f.dayPoint = 1
	f.month = 0
	f.monthPoint = 1
	f.pm = 0
	f.pmPoint = 1
	f.reset()
}

func (f *format) reset() {
	if f.buf == nil {
		f.buf = new(bytes.Buffer)
	} else {
		f.buf.Reset()
	}
}

// Format returns a textual representation of the time value formatted
// according to layout, which defines the format by showing how the reference
// time, defined to be
//	Mon Jan 2 15:04:05 -0700 MST 2006
// would be displayed if it were the value; it serves as an example of the
// desired output. The same display rules will then be applied to the time
// value.
//
// A fractional second is represented by adding a period and zeros
// to the end of the seconds section of layout string, as in "15:04:05.000"
// to format a time stamp with millisecond precision.
//
// Predefined layouts ANSIC, UnixDate, RFC3339 and others describe standard
// and convenient representations of the reference time. For more information
// about the formats and the definition of the reference time, see the
// documentation for ANSIC and the other constants defined by this package.
func (t Time) Format(layout string) string {
	t.f.mixed = false
	return t.lookup(t.Time.Format(t.formatLayout(layout)))
}

// FormatMix the difference with Format is that it can be mixed in English and Chinese.
func (t Time) FormatMix(layout string) string {
	t.f.mixed = true
	return t.lookup(t.Time.Format(t.formatLayout(layout)))
}

// FormatLayout returns a standard textual formatted layout.
// Such as
//	星期一 一月 2 下午15:04:05 -0700 MST 2006
// returns
//	Mon Jan 2 PM15:04:05 -0700 MST 2006
func FormatLayout(layout string) string {
	return new(Time).formatLayout(layout)
}

func (t *Time) formatLayout(layout string) string {
	t.f.init(layout)
	j := 0
	for i, r := range layout {
		if i < j {
			continue
		}
		switch r {
		case 'M':
			if len(layout) >= i+6 && layout[i:i+6] == "Monday" {
				j = i + 6
				if t.f.mixed {
					t.f.day <<= 1
					t.f.dayPoint <<= 1
				}
				t.f.buf.WriteString("Monday")
			} else if len(layout) >= i+3 && layout[i:i+3] == "Mon" {
				j = i + 3
				if t.f.mixed {
					t.f.day <<= 1
					t.f.dayPoint <<= 1
				}
				t.f.buf.WriteString("Mon")
			}
		case 'J':
			if len(layout) >= i+7 && layout[i:i+7] == "January" {
				j = i + 7
				if t.f.mixed {
					t.f.month <<= 1
					t.f.monthPoint <<= 1
				}
				t.f.buf.WriteString("January")
			} else if len(layout) >= i+3 && layout[i:i+3] == "Jan" {
				j = i + 3
				if t.f.mixed {
					t.f.month <<= 1
					t.f.monthPoint <<= 1
				}
				t.f.buf.WriteString("Jan")
			}
		case 'P', 'p':
			if len(layout) >= i+2 {
				if layout[i:i+2] == "PM" || layout[i:i+2] == "pm" {
					j = i + 2
					if t.f.mixed {
						t.f.pm <<= 1
						t.f.pmPoint <<= 1
					}
					t.f.buf.WriteString("PM")
				}
			}
		case '星':
			if len(layout) >= i+9 && layout[i:i+9] == "星期一" {
				j = i + 9
				if t.f.mixed {
					t.f.day = (t.f.day + 1) << 1
					t.f.dayPoint <<= 1
				}
				t.f.buf.WriteString("Monday")
			}
		case '周':
			if len(layout) >= i+6 && layout[i:i+6] == "周一" {
				j = i + 6
				if t.f.mixed {
					t.f.day = (t.f.day + 1) << 1
					t.f.dayPoint <<= 1
				}
				t.f.buf.WriteString("Mon")
			}
		case '一':
			if len(layout) >= i+6 && layout[i:i+6] == "一月" {
				j = i + 6
				if t.f.mixed {
					t.f.month = (t.f.month + 1) << 1
					t.f.monthPoint <<= 1
				}
				t.f.buf.WriteString("Jan")
			}
		case '下':
			if len(layout) >= i+6 && layout[i:i+6] == "下午" {
				j = i + 6
				if t.f.mixed {
					t.f.pm = (t.f.pm + 1) << 1
					t.f.pmPoint <<= 1
				}
				t.f.buf.WriteString("PM")
			}
		}
		if j <= i {
			t.f.buf.WriteRune(r)
		}
	}

	return t.f.buf.String()
}

// FormatValue return a Chinese textual representation of the time value formatted
// according to layout.
func FormatValue(s string) string {
	return new(Time).lookup(s)
}


tables.go
package timezh

var longDayNames = [...]string{
	"Sunday",
	"Monday",
	"Tuesday",
	"Wednesday",
	"Thursday",
	"Friday",
	"Saturday",
}

var shortDayNames = [...]string{
	"Sun",
	"Mon",
	"Tue",
	"Wed",
	"Thu",
	"Fri",
	"Sat",
}

var shortMonthNames = [...]string{
	"Jan",
	"Feb",
	"Mar",
	"Apr",
	"May",
	"Jun",
	"Jul",
	"Aug",
	"Sep",
	"Oct",
	"Nov",
	"Dec",
}

var longMonthNames = [...]string{
	"January",
	"February",
	"March",
	"April",
	"May",
	"June",
	"July",
	"August",
	"September",
	"October",
	"November",
	"December",
}

//var shortPMNames = [...]string{
//	"am",
//	"pm",
//}

var longPMNames = [...]string{
	"AM",
	"PM",
}

// A Weekday specifies a day of the week (Sunday = 0, ...).
const (
	Sunday = iota
	Monday
	Tuesday
	Wednesday
	Thursday
	Friday
	Saturday
)

var longDayMap = [...]string{
	"星期日",
	"星期一",
	"星期二",
	"星期三",
	"星期四",
	"星期五",
	"星期六",
}

var shortDayMap = [...]string{
	"周日",
	"周一",
	"周二",
	"周三",
	"周四",
	"周五",
	"周六",
}

// A Month specifies a month of the year (January = 0, ...).
const (
	January = iota
	February
	March
	April
	May
	June
	July
	August
	September
	October
	November
	December
)

var shortMonthMap = [...]string{
	"一月",
	"二月",
	"三月",
	"四月",
	"五月",
	"六月",
	"七月",
	"八月",
	"九月",
	"十月",
	"十一月",
	"十二月",
}

var longMonthMap = shortMonthMap

// A Day specifies a half of the Day (AM = 0, PM = 0).
const (
	AM = iota
	PM
)

var shortPMMap = [...]string{
	"上午",
	"下午",
}

var longPMMap = shortPMMap


examples/main.go
package main

import (
	"fmt"
	"time"

	"github.com/WindomZ/timezh"
)

func main() {
	t, _ := time.Parse("2006-01-02 15:04:05", "2009-01-03 18:15:05")

	fmt.Println("Time:", t.Format("2006-01-02 15:04:05"))

	format(t)
	parse(t)
}

func format(t time.Time) {
	fmt.Println("===========Format===========")
	fmt.Println("\"time\"效果：")
	fmt.Println("  - time.Format(英文):  ", t.Format("2006-01-02(January) PM3:04:05 Monday(Mon)"))
	fmt.Println("  - time.Format(中文):  ", t.Format("2006年01月02日(January) PM3:04:05 Monday(Mon)"))

	fmt.Println("\"timezh\"同样可以：")
	fmt.Println("  - timezh.Format(英文):", timezh.T(t).Format("2006-01-02(January) PM3:04:05 Monday(Mon)"))
	fmt.Println("  - timezh.Format(中文):", timezh.T(t).Format("2006年01月02日(January) PM3:04:05 Monday(Mon)"))

	fmt.Println("\"timezh\"还可以：")
	fmt.Println("  - timezh.Format(特性):", timezh.T(t).Format("2006年01月02日(一月) 下午3:04:05 星期一(周一)"))
	fmt.Println("  - timezh.Format(特性):", timezh.T(t).Format("2006年01月02日(January, 一月) 下午3:04:05PM 星期一(Mon, 周一)"))
	fmt.Println("  - timezh.Format(特性):", timezh.T(t).FormatMix("2006年01月02日(January, 一月) 下午3:04:05PM 星期一(Mon, 周一)"))
}

func parse(t time.Time) {
	tmp, _ := timezh.Parse("2006年01月02日(一月) 下午3:04:05 星期一(周一)",
		timezh.T(t).Format("2006年01月02日(一月) 下午3:04:05 星期一(周一)"))

	fmt.Println("===========Parse===========")
	fmt.Println("Unix time对比：", t.UnixNano(), "==", tmp.UnixNano())
}
